metrics-influxdb/src/main/java/io/dropwizard/metrics5/influxdb/InfluxDbReporter.java

```diff
@@ -0,0 +1,517 @@
+        writeFieldIfEnabled(MEAN,   snapshot.getMean());
+        writeFieldIfEnabled(MIN,    snapshot.getMin());
+        writeFieldIfEnabled(STDDEV, snapshot.getStdDev());
+        writeFieldIfEnabled(P50,    snapshot.getMedian());
+        writeFieldIfEnabled(P75,    snapshot.get75thPercentile());
+        writeFieldIfEnabled(P95,    snapshot.get95thPercentile());
+        writeFieldIfEnabled(P98,    snapshot.get98thPercentile());
+        writeFieldIfEnabled(P99,    snapshot.get99thPercentile());
+        writeFieldIfEnabled(P999,   snapshot.get999thPercentile());
+        if (hasValues()) {
+            writeTimestampMillis(timestamp);
+            reportLine();
+        }
+    }
+    
+    
+    private void reportMetered(MetricName name, Metered meter, long timestamp) throws IOException {
+        writeMeasurement(name);
+        writeMeteredFieldsIfEnabled(meter);
+        if (hasValues()) {
+            writeTimestampMillis(timestamp);
+            reportLine();
+        }
+    }
+
+    private void writeMeteredFieldsIfEnabled(Metered meter) {
+        writeFieldIfEnabled(COUNT,     meter.getCount());
+        writeFieldIfEnabled(SUM,       meter.getSum());
+        writeFieldIfEnabled(M1_RATE,   convertRate(meter.getOneMinuteRate()));
+        writeFieldIfEnabled(M5_RATE,   convertRate(meter.getFiveMinuteRate()));
+        writeFieldIfEnabled(M15_RATE,  convertRate(meter.getFifteenMinuteRate()));
+        writeFieldIfEnabled(MEAN_RATE, convertRate(meter.getMeanRate()));
+    }
+    
+    private void reportCounter(MetricName name, Counter counter, long timestamp) throws IOException {
+        writeMeasurement(name);
+        writeFieldIfEnabled(COUNT, counter.getCount());
+        if (hasValues()) {
+            writeTimestampMillis(timestamp);
+            reportLine();
+        }
+    }
+
+    private void reportGauge(MetricName name, Gauge<?> gauge, long timestamp) throws IOException {
+        writeMeasurement(name);
+        Object value = gauge.getValue();
+        if (value != null) {
+            writeField(VALUE);
+            if (value instanceof Number) {
+                Number number = (Number) value;
+                if (number instanceof Long || 
+                        number instanceof Integer || 
+                        number instanceof Short || 
+                        number instanceof Byte) {
+                    writeFieldValue(number.longValue());
+                } else {
+                    writeFieldValue(number.doubleValue());
+                }
+            } else if (value instanceof Boolean) {
+                writeFieldValue(((Boolean)value));
+            } else {
+                writeFieldValue(value.toString());
+            }
+        }
+        writeTimestampMillis(timestamp);
+        reportLine();
+    }    
+    
+    private void reset() {
+        str.delete(0, str.length());
+    }
+    
+    private void writeMeasurement(MetricName name) {
```

Comment:
I think it makes to extract all the methods and buffer to a new class like `InfluxDbLineBuilder` or so. Like
```java
class InfluxDbLineBuilder {

    private StringBuilder str = new StringBuilder();
    private boolean firstField;

    private final Set<MetricAttribute> disabledMetricAttributes;
    private final MetricName prefix;
    private final Map<MetricName, String> encodedNameCache = new WeakHashMap<>();

    InfluxDbLineBuffer(Set<MetricAttribute> disabledMetricAttributes, MetricName prefix) {
        this.disabledMetricAttributes = disabledMetricAttributes;
        this.prefix = prefix;
    }

    InfluxDbLineBuffer writeMeasurement(MetricName name) {
        str.delete(0, str.length());
        str.append(encodedNameCache.computeIfAbsent(name, this::writeMeasurementNoCache));
        str.append(' ');
        firstField = true;
        return this;
    }

    public InfluxDbLineBuffer reset() {
        str.delete(0, str.length());
        return this;
    }

    InfluxDbLineBuffer writeField(MetricAttribute key) {
        if (!firstField) {
            str.append(',');
        }
        str.append(key.getCode()).append('=');
        firstField = false;
        return this;
    }

    InfluxDbLineBuffer writeField(String key) {
        if (!firstField) {
            str.append(',');
        }
        appendName(key, str);
        str.append('=');
        firstField = false;
        return this;
    }

    InfluxDbLineBuffer writeFieldValue(double value) {
        str.append(value);
        return this;
    }

    InfluxDbLineBuffer writeFieldValue(long value) {
        str.append(value).append('i');
        return this;
    }

    InfluxDbLineBuffer writeFieldValue(String value) {
        str.append('"');
        appendString(value, str);
        str.append('"');
        return this;
    }

    InfluxDbLineBuffer writeFieldValue(boolean value) {
        str.append(value ? 't' : 'f');
        return this;
    }

    InfluxDbLineBuffer writeTimestampMillis(long utcMillis) {
        str.append(' ').append(utcMillis).append("000000\n");
        return this;
    }

    boolean hasValues() {
        return !firstField;
    }

    InfluxDbLineBuffer writeFieldIfEnabled(MetricAttribute key, double value) {
        if (!disabledMetricAttributes.contains(key)) {
            writeField(key);
            writeFieldValue(value);
        }
        return this;
    }

    InfluxDbLineBuffer writeFieldIfEnabled(MetricAttribute key, long value) {
        if (!disabledMetricAttributes.contains(key)) {
            writeField(key);
            writeFieldValue(value);
        }
        return this;
    }

    public StringBuilder get() {
        return str;
    }
```

Reply:
The idea to not mix the logic of the reporter (reading information about gauges, metrics, timers) from the InfluxDb protocol implementation. The reporter should now know about caches and whether we reuse `StringBuilder` or create a new one. I think it would be nice to see building an InfluxDb metric line as a sequence of calls:
```java
 builder.writeMeasurement(name)
            .writeFieldIfEnabled(COUNT, meter.getCount())
            .writeFieldIfEnabled(SUM, meter.getSum())
            .writeFieldIfEnabled(M1_RATE, convertRate(meter.getOneMinuteRate()))
            .writeFieldIfEnabled(M5_RATE, convertRate(meter.getFiveMinuteRate()))
            .writeFieldIfEnabled(M15_RATE, convertRate(meter.getFifteenMinuteRate()))
            .writeFieldIfEnabled(MEAN_RATE, convertRate(meter.getMeanRate()))
            .writeTimestampMillis(timestamp)
```